examples/raw_rpc_client_subscription.ts

```diff
@@ -1,36 +1,27 @@
 import { assertNotInstanceOf } from "../deps/std/testing/asserts.ts"
 import * as T from "../test_util/mod.ts"
-import * as U from "../util/mod.ts"
 
 const client = await T.polkadot.client
 
-const subscriptionId = await client.subscribe({
-  jsonrpc: "2.0",
-  id: client.providerRef.nextId(),
-  method: "chain_subscribeAllHeads",
-  params: [],
-}, function(e) {
-  assertNotInstanceOf(e, Error)
-  console.log(e)
-  const counter = this.state(U.Counter)
-  if (counter.i === 2) {
-    return this.stop()
-  }
-  counter.inc()
-})
-
-const { result } = U.throwIfError(
-  await client.call({
-    jsonrpc: "2.0",
-    id: client.providerRef.nextId(),
-    method: "chain_unsubscribeAllHeads",
-    params: [subscriptionId],
-  }),
+const result = await client.subscriptionFactory()(
+  "chain_subscribeAllHeads",
+  "chain_unsubscribeNewHeads",
+  [],
+  (ctx) => {
+    let i = 0
+    return (e) => {
+      assertNotInstanceOf(e, Error)
+      console.log(e)
+      if (i === 2) {
+        return ctx.end(true)
+      }
+      i++
+      return
+    }
+  },
 )
 
-console.log(
-  // cspell:disable-next-line
-  `${result ? "S" : "Uns"}uccessfully unsubscribed from subscription ${subscriptionId}`,
-)
+// cspell:disable-next-line
+console.log(`${result ? "S" : "Uns"}uccessfully unsubscribed`)
```

Comment:
Hmm, mightn't `result` be an error, and thus truthy?

Reply:
If there is an error, it will call the listener with the error but ultimately, the promise will resolve to undefined... this is definitely worth revisiting. Hoping @kratico will take a look :)

Reply:
This is related to https://github.com/paritytech/capi/pull/413/files#r1029743739

if there is an error, and the listener does not return it, then the error will be lost.
we need to decide if the listener should handle the error and stop the subscription, or if the client should check for an error, pass it to the listener and instead of returning the listener result, return the error

WDYT?